packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -2822,14 +2823,39 @@ function warnAboutUpdateOnUnmountedFiberInDEV(fiber) {
 
     // If we are currently flushing passive effects, change the warning text.
     if (isFlushingPassiveEffects) {
-      console.error(
-        "Can't perform a React state update from within a useEffect cleanup function. " +
-          'To fix, move state updates to the useEffect() body in %s.%s',
-        tag === ClassComponent
-          ? 'the componentWillUnmount method'
-          : 'a useEffect cleanup function',
-        getStackByFiberInDevAndProd(fiber),
-      );
+      // React can't directly detect a memory leak, but there are some clues that warn about one.
+      // One of these clues is when an unmounted React component tries to update its state.
+      // For example, if a component forgets to remove an event listener when unmounting,
+      // that listener may be called later and try to update state, at which point React would warn about the potential leak.
+      //
+      // Warning signals like this are more useful if they're strong.
+      // For this reason, it's good to always avoid updating state from inside of an effect's cleanup function.
+      // Even when you know there is no potential leak, React has no way to know and so it will warn anyway.
+      // In most cases we suggest moving state updates to the useEffect() body instead.
+      // This works so long as the component is updating its own state (or the state of a descendant).
+      //
+      // However this will not work when a component updates its parent state in a cleanup function.
+      // If such a component is unmounted but its parent remains mounted, the state will be out of sync.
+      // For this reason, we avoid showing the warning if a component is updating an ancestor.
+      let current = fiber;
```

Comment:
Nit: rename to node because “current” without any qualifiers has a specific meaning in other places in code (current tree), and it’s also a bit confusing when compared with “current debug fiber” (it’s hard to remember what both currents mean).